app/src/main/java/com/techcourse/service/UserService.java

```diff
@@ -24,9 +28,27 @@ public void insert(final User user) {
     }
 
     public void changePassword(final long id, final String newPassword, final String createBy) {
-        final var user = findById(id);
-        user.changePassword(newPassword);
-        userDao.update(user);
-        userHistoryDao.log(new UserHistory(user, createBy));
+        final DataSource dataSource = DataSourceConfig.getInstance();
+
+        Connection conn = null;
+        try {
+            conn = dataSource.getConnection();
+            conn.setAutoCommit(false);
+
+            final var user = findById(id);
+            user.changePassword(newPassword);
+            userDao.update(conn, user);
+            userHistoryDao.log(conn, new UserHistory(user, createBy));
+
+            conn.commit();
+        } catch (final SQLException e) {
+            if (conn != null) {
+                try {
+                    conn.rollback();
+                } catch (final SQLException ignored) {
+                }
+            }
+            throw new RuntimeException(e);
+        }
```

Comment:
connection을 명시적으로 `close()` 하거나, try-with-resource로 관리하는 방식은 어떨까요?

```suggestion
        try (final Connection conn = dataSource.getConnection()) {
            try {
                conn.setAutoCommit(false);

                final User user = findById(id);
                user.changePassword(newPassword);
                userDao.update(conn, user);
                userHistoryDao.log(conn, new UserHistory(user, createdBy));

                conn.commit();
            } catch (SQLException e) {
                conn.rollback();
                throw new RuntimeException(e);
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
```

Reply:
오 이런 천재적인 방식이..!
생각도 못했네요 반영했습니다 🫡